projects/plutus/src/ltc_afe.c

```diff
@@ -0,0 +1,310 @@
+#include "crc15.h"
+#include "delay.h"
+#include "ltc_afe.h"
+#include "ltc68041.h"
+#include "plutus_config.h"
+
+static bool s_discharging_cells[LTC6804_CELLS_PER_DEVICE * PLUTUS_AFE_DEVICES_IN_CHAIN] = { false };
+
+static uint16_t s_read_reg_cmd[NUM_LTC_AFE_REGISTER] = {
+  LTC6804_RDCFG_RESERVED,
+  LTC6804_RDCVA_RESERVED,
+  LTC6804_RDCVB_RESERVED,
+  LTC6804_RDCVC_RESERVED,
+  LTC6804_RDCVD_RESERVED,
+  LTC6804_RDAUXA_RESERVED,
+  LTC6804_RDAUXA_RESERVED,
+  LTC6804_RDSTATA_RESERVED,
+  LTC6804_RDSTATB_RESERVED,
+  LTC6804_RDCOMM_RESERVED
+};
+
+static uint8_t s_voltage_reg[NUM_LTC_AFE_VOLTAGE_REGISTER] = {
+  LTC_AFE_REGISTER_CELL_VOLTAGE_A,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_B,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_C,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_D
+};
+
+static void prv_wakeup_idle(const LTCAFESettings *afe) {
+  gpio_set_pin_state(&afe->cs, GPIO_STATE_LOW);
+  delay_us(2);
+  gpio_set_pin_state(&afe->cs, GPIO_STATE_HIGH);
+}
+
+static StatusCode prv_read_register(const LTCAFESettings *afe,
+                                    LTCAFERegister reg, uint8_t *data, size_t len) {
+  if (reg > NUM_LTC_AFE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  uint16_t reg_cmd = s_read_reg_cmd[reg];
+  uint8_t cmd[4] = { 0 };
+
+  cmd[0] = (uint8_t)(reg_cmd >> 8);
+  cmd[1] = (uint8_t)(reg_cmd & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec & 0xFF);
+
+  return spi_exchange(afe->spi_port, cmd, 4, data, len);
+}
+
+static StatusCode prv_read_voltage(LTCAFESettings *afe, LTCAFEVoltageRegister reg, uint8_t *data) {
+  if (reg > NUM_LTC_AFE_VOLTAGE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // 6 bytes in register + 2 bytes for PEC
+  size_t len = ((LTC6804_CELLS_IN_REG * 2) + 2) * PLUTUS_AFE_DEVICES_IN_CHAIN;
+
+  return prv_read_register(afe, s_voltage_reg[reg], data, len);
+}
+
+// start cell voltage conversion
+static void prv_trigger_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADCV command
+  uint16_t adcv = LTC6804_ADCV_RESERVED | LTC6804_ADCV_DISCHARGE_PERMITTED
```

Comment:
We probably want DCP to be low - Otherwise, our readings will be affected by cell balancing.